EIPS/eip-3668.md

```diff
@@ -0,0 +1,184 @@
+
+### Requirement for prefix matching on gateway return data
+The requirement that contracts return a prefix that must match the data returned by the gateway serves as a guard against malicious gateways. Without a required prefix, a gateway can cause the client to call any method on the original contract, or produce a call that validates as correct but answers a different query to the one that the client provided. Allowing the contract to require that the response start with critical information computed from the call data of the original call prevents this attack vector.
+
+More sophisticated conditions could be required - for example, the contract could return a regular expression that must match the returned data. However, this adds significant complexity to the proposal, and due to the nature of ABI encoding, a prefix is sufficient to match a function selector and the first fixed-length arguments of a contract function.
+
+### Use of JSON-RPC
+JSON-RPC is already heavily used in the Ethereum ecosystem as the primary means for communicating with Ethereum nodes. We have adopted it here in order to avoid introducing new RPC specifications and dependencies unnecessarily.
+
+## Backwards Compatibility
+Existing contracts that do not wish to use this specification are unaffected. Clients can add support for Durin to all contract calls without introducing any new overhead or incompatibilities.
+
+Contracts that require Durin will not function in conjunction with clients that do not implement this specification. Attempts to call these contracts from non-compliant clients will result in the contract throwing an exception that is propagaged to the user.
+
+## Reference Implementation
+TBD.
+
+## Security Considerations
+In order to prevent a malicious gateway from causing unintended side-effects or faulty results, contracts MUST specify a `prefix` that is sufficient to ensure the gateway's honesty. For more details, see "Requirement for prefix matching on gateway return data" in the Justification section, and the main specification.
+
+Contracts must also implement sufficient validation of the data returned by the gateway to ensure it is valid. The validation required is application-specific and cannot be specified on a global basis.
+
+Client libraries implementing Durin may want to ensure that callers have to explicitly enable its use on a contract, so users are not surprised when their client library begins making HTTPS calls to external services.
+
+## Glossary
+ - Client: A process, such as JavaScript executing in a web browser, or a backend service, that wishes to query a blockchain for data. The client understands how to fetch data using Durin.
+ - Contract: A smart contract existing on Ethereum or another blockchain.
+ - Gateway: A service that answers application-specific Durin queries, usually over HTTPS.
+
```

Comment:
```suggestion
```
This isn't a valid top level EIP section.  Consider folding this into one of the standard sections like the specification (if these terms are referenced in the specification, this **SHOULD** be part of the specification).

Reply:
EIP 1 specifies which sections an EIP *should* have. It doesn't say it's an exhaustive list!

If you strongly feel the glossary should be part of the specification section, I can move it - but that's not where glossaries are commonly put.

Reply:
For a while now we have been enforcing that sections match the template because people were creating all sorts of random sections and EIP documents were becoming quite non-standard looking.

To get ahead of your next question: yes, we should update EIP-1 to match the effective rules.  However, I have been lazy about that because the *only* people who ever refer to EIP-1 when I suggest things are ex-EIP editors (you, Alex, Greg).  I don't think any regular users actually read EIP-1 so updating it would basically be a formality to get you three to do what everyone else does.  😄  I recognize that this is very much not an ideal situation, and I have been trying (along with others) to get more EIP-editors so we can tackle these problems but finding editors is hard.  😢

Reply:
I'm skeptical that enforcing this restriction is actually useful. Would this spec be more readable if the glossary was a subheading of 'specification' to be technically compliant with the unwritten rules? This seems an awful lot like enforcing a rule because it's there rather than because it improves anything.

Reply:
It is useful when you are optimizing for minimum amount of EIP Editor effort rather than maximum EIP quality, which is the current state of things.  At the moment, the *actual* options on the table are:
1. Strict but very easy/low cost to enforce rules that are uniformly applied.
2. No EIP editing happens at all.

It would be *great* if we could get to a more healthy place with regards to EIP editing where we had enough editor-hours available to be more thoughtful and nuanced in the review process, but unfortunately that isn't the world we live in right now.

The specific problem here is that if I say OK to you, I'll soon have people saying "but you let Nick submit an EIP with a non-standard section, so why can't I" (yes, this actually does happen when exceptions are made) and then I have to have a likely long nuanced debate with an author about why Nick's extraneous section was acceptable while theirs was not.  I think such conversations would be very educational and valuable for everyone involved, if we had an unbounded amount of editor-hours available to have them, but we do not sadly.